### ISSUE TITLE
concat with sb doesn't support bdev_raid_add_base_bdev

### ISSUE THREAD

**Issue by clay-mayers** [spdk/spdk#3535]
# Sighting report

I am hot unplugging and hot plugging back in the same device, but concat doesn't "rebuild" itself because it seems to have executed bdev_raid_remove_base_bdev when the device was removed (I see a null uuid for the removed base bdev).   To compensate, I'm trying to use bdev_raid_add_base_bdev to add the device back in, but it fails.

## Expected Behavior

I expect concat to support having a base device removed with bdev_raid_remove_base_bdev and replacing it with a new or previous base bdev with bdev_raid_add_base_bdev.

## Current Behavior

bdev_raid_add_base_bdev errors with `*ERROR*: Bdev malloc0 is not a member of raid bdev concat0` 

## Steps to Reproduce

Using spdk_tgt as the spdk app with these rpc.py commands, it will fail on line 5

1. bdev_malloc_create -b malloc0 64 512
2. bdev_malloc_create -b malloc1 64 512
3. bdev_raid_create -n concat0 -r concat -z 1 -s -b "malloc0 malloc1"
4. bdev_raid_remove_base_bdev malloc0
5. bdev_raid_add_base_bdev concat0 malloc0

## Context (Environment including OS version, SPDK version, etc.)

Running on `5.15.0-122-generic #132-Ubuntu SMP Thu Aug 29 13:45:52 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux` 
with branch 24.05.x at commit 29f97811d5b (May 24th 2024). 

```
$ sudo scripts/rpc.py bdev_malloc_create -b malloc0 64 512
malloc0
$ sudo scripts/rpc.py bdev_malloc_create -b malloc1 64 512
malloc1
$ sudo scripts/rpc.py bdev_raid_create -n concat0 -r concat -z 1 -s -b "malloc0 malloc1"
$ sudo scripts/rpc.py bdev_raid_remove_base_bdev malloc0
$ sudo scripts/rpc.py bdev_raid_add_base_bdev concat0 malloc0
[2024-09-24 23:10:02.850746] bdev_raid.c:3663:raid_bdev_examine_sb: *ERROR*: Bdev malloc0 is not a member of raid bdev concat0
request:
{
  "raid_bdev": "concat0",
  "base_bdev": "malloc0",
  "method": "bdev_raid_add_base_bdev",
  "req_id": 1
}
Got JSON-RPC error response
response:
{
  "code": -22,
  "message": "Failed to add base bdev to RAID bdev: Invalid argument"
}
```

**Comment by jimharris** [comment]
[Bug scrub] @apaszkie, can you take a look?

**Comment by apaszkie** [comment]
When a base bdev is removed from a concat (or raid0) raid bdev the array state changes to "offline" and the bdev is not usable anymore. That's why it doesn't allow to re-add the base bdev. It is not possible to rebuild raid arrays without redundancy, similarly to removing more than one disk from raid5.

**Comment by clay-mayers** [comment]
It goes offline because of the lack of redundancy, but shouldn't it go back online when the base bdev is set and the min required devices has been restored?   Sure the RAID device can't restore the user data but shouldn't it restore the device itself (there is redundancy of the md with the sb)?  Data recovery becomes the client's responsibility in this case.

**Comment by apaszkie** [comment]
I'm not sure about allowing to bring it back from "offline" state. That's essentially resurrecting a failed array. To make the behavior consistent it would have to be allowed also for raid5f and others. Interestingly, there is another [issue](https://github.com/spdk/spdk/issues/3497) where the expectation is that we _don't_ allow restoring from offline state. Maybe as a compromise we could allow it when passing some option (like "--force") to bdev_raid_add_base_bdev?

**Comment by jimharris** [comment]
I think RAID-0 and concat are fundamentally different than RAID-5F or RAID-1.

RAID-5F and RAID-1 have rebuild characteristics involved. If a drive gets removed, the RAID volume itself still remains functional, so when you put the original disk back in, the RAID volume can't be restored back to its original state.

But if a RAID-0 or concat disk gets removed, the RAID volume is immediately failed. Plugging it back in immediately makes the RAID volume operational again.

**Comment by apaszkie** [comment]
True, but removing two drives from raid5f is the same as removing one or more from raid0 or concat - the array is failed. Should we allow to add a drive to a failed raid5f array?

**Comment by jimharris** [comment]
> True, but removing two drives from raid5f is the same as removing one or more from raid0 or concat - the array is failed. Should we allow to add a drive to a failed raid5f array?

No, after a drive is removed from a failed raid1 or raid5f array, it can't be added back in as-is.

**Comment by clay-mayers** [comment]
I've tried deleting the concat device and recreating it but 24.05.x seg faults checking bdev->sb->seq_number as sb is null.
```
bdev_malloc_create -b malloc0 64 512
bdev_malloc_create -b malloc1 64 512
bdev_raid_create -n concat0 -r concat -z 1 -s -b "malloc0 malloc1" --uuid ce204084-b3cd-484d-a351-3e627fed3ab2
bdev_raid_remove_base_bdev malloc0
bdev_raid_delete concat0
bdev_raid_create -n concat0 -r concat -z 1 -s -b "malloc0 malloc1" --uuid ce204084-b3cd-484d-a351-3e627fed3ab2
```
Produces this
```
malloc0
malloc1
Exception:
bdev_malloc_create -b malloc0 64 512
bdev_malloc_create -b malloc1 64 512
bdev_raid_create -n concat0 -r concat -z 1 -s -b "malloc0 malloc1" --uuid ce204084-b3cd-484d-a351-3e627fed3ab2
bdev_raid_remove_base_bdev malloc0
bdev_raid_delete concat0
bdev_raid_create -n concat0 -r concat -z 1 -s -b "malloc0 malloc1" --uuid ce204084-b3cd-484d-a351-3e627fed3ab2 <<<
Connection closed with partial response:

[1]+  Segmentation fault      sudo build/bin/spdk_tgt -m[0-3]
```

**Comment by clay-mayers** [comment]
My actual test case is pcie drive removal (using QEMU) - I can get everything restored by deleting concat, then plugging the device back in.   It seems, what's really in my way is that concat doesn't delete itself and just sits in a failed state - yet all the devices (lvstore/lvols) above it are gone as far as I can tell.

Why is the concat device kept around if it's not useful for anything?
